### ISSUE TITLE
Librosa 0.7.1 load() gives Malformed 'fmt ' chunk (worked in 0.7.0)

### ISSUE THREAD

**Issue by NumesSanguis** [librosa/librosa#1019]
#### Description

In v0.7.0 company_sensitive.wav worked, but with v0.7.1 throws an "Malformed 'fmt ' chunk." error.
Unfortunately I'm not able to provide said file, but I'll try to find another one / remove sensitive data later.

#### Steps/Code to Reproduce
```
import librosa

aud, sr = librosa.load(Path("/egg/foo.wav"), sr=None, mono=False)
```

#### Expected Results
No output, and `aud` to contain numpy formatted audio data.

#### Actual Results
```python
Error opening '/egg/foo.wav': Error in WAV/W64/RF64 file. Malformed 'fmt ' chunk.
Traceback (most recent call last):
  File "/opt/conda/lib/python3.7/site-packages/airflow/models/taskinstance.py", line 922, in _run_raw_task
    result = task_copy.execute(context=context)
  File "/opt/conda/lib/python3.7/site-packages/airflow/operators/python_operator.py", line 113, in execute
    return_value = self.execute_callable()
  File "/opt/conda/lib/python3.7/site-packages/airflow/operators/python_operator.py", line 118, in execute_callable
    return self.python_callable(*self.op_args, **self.op_kwargs)
  File "/usr/local/airflow/dags/scripts_processing/*company*/audio_parser.py", line 187, in start
    return task_class.audio_parser(kwargs['params']["expected_sr"])
  File "/usr/local/airflow/dags/scripts_processing/*company*/audio_parser.py", line 65, in audio_parser
    audio_concat, file_metadata = self.combine_audio(audio_folder[0], sr_expected=expected_sr)
  File "/usr/local/airflow/dags/scripts_processing/*company*/audio_parser.py", line 109, in combine_audio
    aud, sr = librosa.load(audio_file, sr=None, mono=False)
  File "/opt/conda/lib/python3.7/site-packages/librosa/core/audio.py", line 149, in load
    six.reraise(*sys.exc_info())
  File "/opt/conda/lib/python3.7/site-packages/six.py", line 696, in reraise
    raise value
  File "/opt/conda/lib/python3.7/site-packages/librosa/core/audio.py", line 129, in load
    with sf.SoundFile(path) as sf_desc:
  File "/opt/conda/lib/python3.7/site-packages/soundfile.py", line 627, in __init__
    self._file = self._open(file, mode_int, closefd)
  File "/opt/conda/lib/python3.7/site-packages/soundfile.py", line 1182, in _open
    "Error opening {0!r}: ".format(self.name))
  File "/opt/conda/lib/python3.7/site-packages/soundfile.py", line 1355, in _error_check
    raise RuntimeError(prefix + _ffi.string(err_str).decode('utf-8', 'replace'))
RuntimeError: Error opening '/egg/foo.wav': Error in WAV/W64/RF64 file. Malformed 'fmt ' chunk.
```

#### Versions
Using Anaconda

##### Broken v0.7.1
Linux-4.15.0-65-generic-x86_64-with-debian-buster-sid
Python 3.7.5 (default, Oct 25 2019, 15:51:11) 
[GCC 7.3.0]
NumPy 1.17.3
SciPy 1.3.2
librosa 0.7.1

##### Working v0.7.0
Linux-4.15.0-65-generic-x86_64-with-debian-buster-sid
Python 3.7.5 (default, Oct 25 2019, 15:51:11) 
[GCC 7.3.0]
NumPy 1.17.3
SciPy 1.3.1
librosa 0.7.0

**Comment by lostanlen** [comment]
Hello @NumesSanguis. Note that the error arises in PySoundFile. You said that the `load` passes on librosa 0.7.0 but fails on librosa 0.7.1? Can you please make sure that the versions of PySoundFile were the same in both cases? If not, then i would suggest digging further in that direction. If so, then it's a bit surprising, given that we didn't make any major change to our pysoundfile call sites between 0.7.0 and 0.7.1. (https://github.com/librosa/librosa/pull/968 perhaps?)

TL;DR please report your PySoundFile versions

**Comment by lostanlen** [comment]
Worth reading: https://github.com/supercollider/supercollider/issues/3137

**Comment by NumesSanguis** [comment]
Output of:
```
import soundfile as sf
print(sf.__version__)
```
Failing soundfile version: `0.10.2`
Working soundfile version: `0.10.2`

I'm also surprised by this result. Originally I thought it was related to the change made `0.7.0` to make PySoundFile the default loader, but when checking my versions, I was already on `0.7.0`.

I inquired in my company about this data, and it turned out the file is actually a `mp3`, but they renamed it `.wav` :')
Through your link, I could indeed confirm this with `file /egg/foo.wav`:
```
RIFF (little-endian) data, WAVE audio, MPEG Layer 3, stereo 8000 Hz
```
Weird that it does load in `0.7.0` though, but not in `0.7.1`

Did something change in the fallback code or something?
I see here there that in case of a `RunTimeError` Librosa falls back to `audioread`. Maybe it's an idea to also try this in case of error such as my `Malformed 'fmt ' chunk`?

**Comment by NumesSanguis** [comment]
The following steps reproduce the error in `0.7.1`, but not in `0.7.0`

1. Download a sample .mp3 file to your current folder: `curl https://file-examples.com/wp-content/uploads/2017/11/file_example_MP3_700KB.mp3 --output sample.mp3`
2. Create 2 conda envs with version `0.7.0` and `0.7.1` (2x):
```
conda create --name librosa070 python=3.7  # librosa071
conda activate librosa070  # librosa071
conda install -c conda-forge librosa=0.7.0  # 0.7.1
```
3. Creation a python file `librosa_regression.py` with the following code:
```python
from pathlib import Path
import librosa

import soundfile as sf
print("PySoundFile", sf.__version__)

import platform; print(platform.platform())
import sys; print("Python", sys.version)
import numpy; print("NumPy", numpy.__version__)
import scipy; print("SciPy", scipy.__version__)
import librosa; print("librosa", librosa.__version__)

aud_file = "sample.mp3"
aud, sr = librosa.load(Path(aud_file), sr=None, mono=False)
print(aud.shape)
```
4. Librosa `v0.7.0`: **Works**
```bash
$ python librosa_regression.py 
PySoundFile 0.10.2
Linux-4.15.0-70-generic-x86_64-with-debian-buster-sid
Python 3.7.5 (default, Oct 25 2019, 15:51:11) 
[GCC 7.3.0]
NumPy 1.17.3
SciPy 1.3.2
librosa 0.7.0
(2, 870191)
```
5. Librosa `v0.7.1`: **Broken**
```bash
$ python librosa_regression.py 
PySoundFile 0.10.2
Linux-4.15.0-70-generic-x86_64-with-debian-buster-sid
Python 3.7.5 (default, Oct 25 2019, 15:51:11) 
[GCC 7.3.0]
NumPy 1.17.3
SciPy 1.3.2
librosa 0.7.1
Traceback (most recent call last):
  File "librosa_regression.py", line 14, in <module>
    aud, sr = librosa.load(Path(aud_file), sr=None, mono=False)
  File "/home/*user*/anaconda3/envs/librosa071/lib/python3.7/site-packages/librosa/core/audio.py", line 149, in load
    six.reraise(*sys.exc_info())
  File "/home/*user*/anaconda3/envs/librosa071/lib/python3.7/site-packages/six.py", line 696, in reraise
    raise value
  File "/home/*user*/anaconda3/envs/librosa071/lib/python3.7/site-packages/librosa/core/audio.py", line 129, in load
    with sf.SoundFile(path) as sf_desc:
  File "/home/*user*/anaconda3/envs/librosa071/lib/python3.7/site-packages/soundfile.py", line 627, in __init__
    self._file = self._open(file, mode_int, closefd)
  File "/home/*user*/anaconda3/envs/librosa071/lib/python3.7/site-packages/soundfile.py", line 1182, in _open
    "Error opening {0!r}: ".format(self.name))
  File "/home/*user*/anaconda3/envs/librosa071/lib/python3.7/site-packages/soundfile.py", line 1355, in _error_check
    raise RuntimeError(prefix + _ffi.string(err_str).decode('utf-8', 'replace'))
RuntimeError: Error opening 'sample.mp3': File contains data in an unknown format.
```

**Comment by lostanlen** [comment]
Thank you for such a detailed error report, and thank you for making a minimal script to reproduce the problem.
I ran `librosa_regression.py` and found an error in both cases.

```
(librosa070) 4c32759bd91d:Downloads vl238$ python librosa_regression.py
PySoundFile 0.10.2
Darwin-18.7.0-x86_64-i386-64bit
Python 3.7.5 (default, Oct 25 2019, 10:52:18)
[Clang 4.0.1 (tags/RELEASE_401/final)]
NumPy 1.17.3
SciPy 1.3.1
librosa 0.7.0
Traceback (most recent call last):
  File "/Users/vl238/miniconda3/envs/librosa070/lib/python3.7/site-packages/librosa/core/audio.py", line 127, in load
    with sf.SoundFile(path) as sf_desc:
  File "/Users/vl238/miniconda3/envs/librosa070/lib/python3.7/site-packages/soundfile.py", line 627, in __init__
    self._file = self._open(file, mode_int, closefd)
  File "/Users/vl238/miniconda3/envs/librosa070/lib/python3.7/site-packages/soundfile.py", line 1182, in _open
    "Error opening {0!r}: ".format(self.name))
  File "/Users/vl238/miniconda3/envs/librosa070/lib/python3.7/site-packages/soundfile.py", line 1355, in _error_check
    raise RuntimeError(prefix + _ffi.string(err_str).decode('utf-8', 'replace'))
RuntimeError: Error opening 'sample.mp3': File contains data in an unknown format.
```

Note that this is on a Mac OS X (Darwin-18.7.0-x86_64-i386-64bit), not Linux.
The plot thickens

**Comment by bmcfee** [comment]
> Did something change in the fallback code or something?
> I see here there that in case of a `RunTimeError` Librosa falls back to `audioread`. Maybe it's an idea to also try this in case of error such as my `Malformed 'fmt ' chunk`?

I think it would be best to leave it as a failure.  For better or worse, the error message here was diagnostic and correct, and burying it under a try-catch to use a more permissive decoder would have obscured the problem.

There were a few changes to load exception handling between 0.7.0 and 0.7.1, but I think the current behavior is "more correct" in the sense described above, so I'm not inclined to go change it.  You can see a full breakdown here: https://github.com/librosa/librosa/commits/master/librosa/core/audio.py ; 0.7.0 was tagged on July 7, 2019, so anything after that could be relevant.

**Comment by NumesSanguis** [comment]
@bmcfee Are you arguing for that this checking should stay like this, and instead of rolling back the change, Librosa should provide way to deal with non-wav formatted audio (e.g. user should specify `engine='audioread'`. Or do you mean that Librosa will keep `.load()` like this, not being able to load non-wav formatted data?

In case of the 2nd interpretation, I'm not disagreeing that this behavior is "more correct", but it would be a reason for me to switch to another library. Librosa has been for me the no-nonsense, easy to use audio library. If we now have to jump through loops to get non-wav audio to load, from a user-perspective, it would be a regression in the usability of this library.

TL;DR How should the user load non-correctly wav formatted audio?

**Comment by bmcfee** [comment]
Maybe I didn't understand correctly, but I thought the problem was that your audio file had the wrong extension, which caused sndfile to fail because it was using the wrong codec?  This seems like correct behavior to me, as it identified a problem in your data. If the data isn't correctly formatted, I don't think it's reasonable to expect it to decode, even if some libraries allow it to pass through silently.

My suggested solution is to use the error messages to identify badly formatted files and correct the data.

**Comment by NumesSanguis** [comment]
@bmcfee Ah yes, in my company's case it was indeed a wrong extension. However, the minimal example I've given above, the file downloaded is a `.mp3` and it still fails (`RuntimeError: Error opening 'sample.mp3': File contains data in an unknown format.`). So even with the correct file extension, the same error still occurs (works in `0.7.0`).

Also, I renamed my company's file to `.mp3` and I still get the error in `0.7.1`: `RuntimeError: Error opening 'company_sensitive.mp3': Error in WAV/W64/RF64 file. Malformed 'fmt ' chunk.`

<details><summary>Full error log (CLICK ME)</summary>
<p>

```python
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-1-53f038775bd6> in <module>
      8 wav_file = "company_sensitive.mp3"
      9 # wav_file = "sample.mp3"  # "test.wav"
---> 10 aud, sr = librosa.load(Path(wav_file), sr=None, mono=False)
     11 print(aud.shape)
     12 print(aud.dtype)

/opt/miniconda3/envs/audio_tester/lib/python3.7/site-packages/librosa/core/audio.py in load(path, sr, mono, offset, duration, dtype, res_type)
    147             y, sr_native = __audioread_load(path, offset, duration, dtype)
    148         else:
--> 149             six.reraise(*sys.exc_info())
    150 
    151     # Final cleanup for dtype and contiguity

/opt/miniconda3/envs/audio_tester/lib/python3.7/site-packages/six.py in reraise(tp, value, tb)
    694             if value.__traceback__ is not tb:
    695                 raise value.with_traceback(tb)
--> 696             raise value
    697         finally:
    698             value = None

/opt/miniconda3/envs/audio_tester/lib/python3.7/site-packages/librosa/core/audio.py in load(path, sr, mono, offset, duration, dtype, res_type)
    127 
    128     try:
--> 129         with sf.SoundFile(path) as sf_desc:
    130             sr_native = sf_desc.samplerate
    131             if offset:

/opt/miniconda3/envs/audio_tester/lib/python3.7/site-packages/soundfile.py in __init__(self, file, mode, samplerate, channels, subtype, endian, format, closefd)
    625         self._info = _create_info_struct(file, mode, samplerate, channels,
    626                                          format, subtype, endian)
--> 627         self._file = self._open(file, mode_int, closefd)
    628         if set(mode).issuperset('r+') and self.seekable():
    629             # Move write position to 0 (like in Python file objects)

/opt/miniconda3/envs/audio_tester/lib/python3.7/site-packages/soundfile.py in _open(self, file, mode_int, closefd)
   1180             raise TypeError("Invalid file: {0!r}".format(self.name))
   1181         _error_check(_snd.sf_error(file_ptr),
-> 1182                      "Error opening {0!r}: ".format(self.name))
   1183         if mode_int == _snd.SFM_WRITE:
   1184             # Due to a bug in libsndfile version <= 1.0.25, frames != 0

/opt/miniconda3/envs/audio_tester/lib/python3.7/site-packages/soundfile.py in _error_check(err, prefix)
   1353     if err != 0:
   1354         err_str = _snd.sf_error_number(err)
-> 1355         raise RuntimeError(prefix + _ffi.string(err_str).decode('utf-8', 'replace'))
   1356 
   1357 

RuntimeError: Error opening 'company_sensitive.mp3': Error in WAV/W64/RF64 file. Malformed 'fmt ' chunk.
```

</p>
</details>

So the change from `0.7.0` to `0.7.1` is effectively dropped support for `.mp3` / `RIFF`?

**Comment by bmcfee** [comment]
No, we did not drop support for mp3. Are you absolutely sure that there data is well-formed? If it's mp3, then pysoundfile should not be decoding it.

At any rate, we're not really in the decoding game here. All decoding is done by either pysoundfile or audioread, and the logic on our side hasn't fundamentally changed except to be more careful about error handling.

**Comment by NumesSanguis** [comment]
@bmcfee Please refer to my minimal example above: https://github.com/librosa/librosa/issues/1019#issuecomment-554819300
It downloads a public .mp3, which you can test yourself.

Somehow this error handling has made it for 0.7.1 impossible to load certain kinds of audio data, which was possible for 0.7.0.

**Comment by bmcfee** [comment]
I tried your example, and it appears to be platform dependent.  On my system (ubuntu 19.04, conda py3.7), it loads as expected, with a warning when pysoundfile fails and a fallback to audioread.  I believe audioread is then calling out to ffmpeg to do the decoding:

```python
In [1]: import librosa                                                                        
y, sr = libr
In [2]: y, sr = librosa.load('sample.mp3')                                                    
/home/bmcfee/git/librosa/librosa/core/audio.py:146: UserWarning: PySoundFile failed. Trying audioread instead.
  warnings.warn('PySoundFile failed. Trying audioread instead.')

In [3]: y, sr                                                                                 
Out[3]: 
(array([ 0.        ,  0.        ,  0.        , ..., -0.00077047,
        -0.0006783 ,  0.        ], dtype=float32), 22050)

In [4]: librosa.show_versions()                                                               
INSTALLED VERSIONS
------------------
python: 3.7.5 (default, Oct 25 2019, 15:51:11) 
[GCC 7.3.0]

librosa: 0.7.1

audioread: 2.1.8
numpy: 1.17.3
scipy: 1.3.0
sklearn: 0.21.2
joblib: 0.13.2
decorator: 4.4.0
six: 1.13.0
soundfile: 0.10.2
resampy: 0.2.2
numba: 0.44.1

numpydoc: 0.9.1
sphinx: 2.2.1
sphinx_rtd_theme: 0.4.2
sphinxcontrib.versioning: 2.2.1
sphinx-gallery: None
pytest: 3.10.1
pytest-mpl: None
pytest-cov: None
matplotlib: 3.1.1
presets: 0.1.3
```

For full context, here are the only changes to `load` from 0.7.0 to my current dev branch (equivalent to 0.7.1 for this file):

```diff
-        # If soundfile failed, fall back to the audioread loader
-        y, sr_native = __audioread_load(path, offset, duration, dtype)
+
+        # If soundfile failed, try audioread instead
+        if isinstance(path, six.string_types):
+            warnings.warn('PySoundFile failed. Trying audioread instead.')
+            y, sr_native = __audioread_load(path, offset, duration, dtype)
+        else:
+            six.reraise(*sys.exc_info())
```
The only functional difference here is that audioread fallback is only triggered if the input to load is a string-type (ie a file path).  This was implemented in #load

**Comment by bmcfee** [comment]
Aha!  I now see the problem.  @NumesSanguis you appear to be using pathlib to manage file paths, and we're making an explicit check for string types (which excludes pathlib).  This is the source of the failure, and not anything to do with the underlying format support.

For now, I suggest to cast your path to a string-type and everything should right itself.  Going forward, we should relax this check to support pathlib, though I think we can't properly do that until python 2.7 support is formally dropped in the 0.8 series.

**Comment by NumesSanguis** [comment]
@bmcfee It was indeed pathlib!

Since the error is not very helpful to find the source of the problem, and it is better to not push the responsibility to fix it on the user, I added a converter from pathlib object to String.
Basically it comes down to:
Python 3.6+: `path = os.fspath(path)`
Python 3.4+: `path = str(path)`

When Librosa 0.8.x comes around, this temporary fix can be removed, and only this has to change (also mentioned in the pull request):
```python
import pathlib

# if isinstance(path, six.string_types):  -->
if isinstance(path, (six.string_types, pathlib.PurePath)):
```
